src/tools/rbd_mirror/group_replayer/BootstrapRequest.cc

```diff
@@ -0,0 +1,1469 @@
+
+namespace {
+
+static const uint32_t MAX_RETURN = 1024;
+
+bool is_demoted_snap_exists(
+    const std::vector<cls::rbd::GroupSnapshot> &snaps) {
+  for (auto it = snaps.rbegin(); it != snaps.rend(); it++) {
+     auto ns = std::get_if<cls::rbd::MirrorGroupSnapshotNamespace>(
+       &it->snapshot_namespace);
+    if (ns != nullptr) {
+      if (ns->is_demoted()) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+int get_last_mirror_snapshot_state(
+    const std::vector<cls::rbd::GroupSnapshot> &snaps,
+    cls::rbd::MirrorSnapshotState *state) {
+  for (auto it = snaps.rbegin(); it != snaps.rend(); it++) {
+    auto ns = std::get_if<cls::rbd::MirrorGroupSnapshotNamespace>(
+        &it->snapshot_namespace);
+    if (ns != nullptr) {
+      // XXXMG: check primary_mirror_uuid matches?
+      *state = ns->state;
+      return 0;
+    }
+  }
+
+  return -ENOENT;
+}
+
+} // anonymous namespace
+
+template <typename I>
+BootstrapRequest<I>::BootstrapRequest(
+    Threads<I> *threads,
+    librados::IoCtx &local_io_ctx,
+    librados::IoCtx &remote_io_ctx,
+    const std::string &global_group_id,
+    const std::string &local_mirror_uuid,
+    InstanceWatcher<I> *instance_watcher,
+    MirrorStatusUpdater<I> *local_status_updater,
+    MirrorStatusUpdater<I> *remote_status_updater,
+    journal::CacheManagerHandler *cache_manager_handler,
+    PoolMetaCache *pool_meta_cache,
+    bool *resync_requested,
+    std::string *local_group_id,
+    std::string *remote_group_id,
+    std::map<std::string, cls::rbd::GroupSnapshot> *local_group_snaps,
+    GroupCtx *local_group_ctx,
+    std::list<std::pair<librados::IoCtx, ImageReplayer<I> *>> *image_replayers,
+    std::map<std::pair<int64_t, std::string>, ImageReplayer<I> *> *image_replayer_index,
+    Context* on_finish)
+  : CancelableRequest("rbd::mirror::group_replayer::BootstrapRequest",
+		      reinterpret_cast<CephContext*>(local_io_ctx.cct()),
+                      on_finish),
+    m_threads(threads),
+    m_local_io_ctx(local_io_ctx),
+    m_remote_io_ctx(remote_io_ctx),
+    m_global_group_id(global_group_id),
+    m_local_mirror_uuid(local_mirror_uuid),
+    m_instance_watcher(instance_watcher),
+    m_local_status_updater(local_status_updater),
+    m_remote_status_updater(remote_status_updater),
+    m_cache_manager_handler(cache_manager_handler),
+    m_pool_meta_cache(pool_meta_cache),
+    m_resync_requested(resync_requested),
+    m_local_group_id(local_group_id),
+    m_remote_group_id(remote_group_id),
+    m_local_group_snaps(local_group_snaps),
+    m_local_group_ctx(local_group_ctx),
+    m_image_replayers(image_replayers),
+    m_image_replayer_index(image_replayer_index),
+    m_on_finish(on_finish) {
+  dout(10)  << "global_group_id=" << m_global_group_id << dendl;
+}
+
+template <typename I>
+void BootstrapRequest<I>::send() {
+  *m_resync_requested = false;
+
+  std::string group_id;
+
+  if (m_local_group_id && !m_local_group_id->empty()) {
```

Comment:
When is this likely to not be empty? The BootstrapRequest is called when the GroupReplayer is first created and hence has no local group information. It is the BootstrapRequest that gets this information so the oid will never be correct at this point.

Reply:
Ok, got it. This will be called again when the Groupreplayer is restarted.